### PR TITLE
=core Make use of statefulMap instead of statefulMapConcat.

### PR DESCRIPTION
**Pull request by He-Pin**
Motivation:
Use the `statefulMap` instead when concating is not needed.

```
[info] Benchmark                                                Mode  Cnt         Score         Error  Units
[info] MessageToFrameRendererBenchmark.benchNewStreamedFrames  thrpt   10  25333318.747 锟斤拷 1293108.754  ops/s
[info] MessageToFrameRendererBenchmark.benchOldStreamedFrames  thrpt   10  20352150.287 锟斤拷 1563364.567  ops/s
```

### CODE REVIEW

File: http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala
Comment (pjfanning) [inline http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala:267]: Can you deprecate this? We can't remove public methods without deprecating them first.

Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala:267]: The StreamUtil is private[http].

**Review by pjfanning — CHANGES_REQUESTED** [review]
could you have a look at #312 ? I'm going to block this until we know if there is a mima failure

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/ws/MessageToFrameRenderer.scala
Comment (jrudolph) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/ws/MessageToFrameRenderer.scala:39]: How much more boxing and allocations will that produce?

File: http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala
Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala:267]: image

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/ws/MessageToFrameRenderer.scala
Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/impl/engine/ws/MessageToFrameRenderer.scala:39]:
```
[info] Benchmark                                                Mode  Cnt         Score        Error  Units
[info] MessageToFrameRendererBenchmark.benchNewStreamedFrames  thrpt   10  15148931.446 锟斤拷  99406.646  ops/s
[info] MessageToFrameRendererBenchmark.benchOldStreamedFrames  thrpt   10  13296638.654 锟斤拷 553890.759  ops/s
```
image

**Comment by He-Pin** [comment]
@jrudolph @pjfanning @mdedetrich ping~

File: http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala
Comment (pjfanning) [inline http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala:1]: should this file have a Lightbend header?

**Comment by pjfanning** [comment]
@He-Pin can you rebase this PR - it's quite old

File: http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala
Comment (He-Pin) [reply http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala:1]: Let me check

Comment (pjfanning) [reply http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala:1]: @He-Pin I will approve this PR if we can work out what license header is needed here.

Comment (He-Pin) [reply http-bench-jmh/src/main/scala/org/apache/pekko/BenchTestSource.scala:1]: Yes, I copied it from pekko, where it has a lightbend header.

**Review by pjfanning — APPROVED** [review]
lgtm

**Pull request merged**
